runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp

```diff
@@ -0,0 +1,1529 @@
+				/* Drop arrayObject */
+				DROP_OBJECT_IN_SPECIAL_FRAME(currentThread);
+				vmFuncs->setHeapOutOfMemoryError(currentThread);
+			} else {
+				arrayObject = POP_OBJECT_IN_SPECIAL_FRAME(currentThread);
+				j9object_t membernameObject = J9_JNI_UNWRAP_REFERENCE(self);
+				jint flags = J9VMJAVALANGINVOKEMEMBERNAME_FLAGS(currentThread, membernameObject);
+				jlong vmindex = (jlong)(UDATA)J9OBJECT_U64_LOAD(currentThread, membernameObject, vm->vmindexOffset);
+				j9object_t target;
+				if (J9_ARE_ANY_BITS_SET(flags, MN_IS_FIELD)) {
+					vmindex = ((J9JNIFieldID*)vmindex)->offset;
+					target = J9VMJAVALANGINVOKEMEMBERNAME_CLAZZ(currentThread, membernameObject);
+				} else {
+					J9JNIMethodID *methodID = (J9JNIMethodID*)vmindex;
+					if (J9_ARE_ANY_BITS_SET(methodID->vTableIndex, J9_JNI_MID_INTERFACE)) {
+						vmindex = methodID->vTableIndex & ~J9_JNI_MID_INTERFACE;
+					} else if (0 == methodID->vTableIndex) {
+						vmindex = -1;
+					} else {
+						vmindex = methodID->vTableIndex;
+					}
+					target = membernameObject;
+				}
+
+				J9VMJAVALANGLONG_SET_VALUE(currentThread, box, vmindex);
+				J9JAVAARRAYOFOBJECT_STORE(currentThread, arrayObject, 0, box);
+				J9JAVAARRAYOFOBJECT_STORE(currentThread, arrayObject, 0, target);
+
+				result = vmFuncs->j9jni_createLocalRef(env, arrayObject);
+			}
+		}
+	}
+	Trc_JCL_java_lang_invoke_MethodHandleNatives_getMemberVMInfo_Exit(env, result);
+	vmFuncs->internalExitVMToJNI(currentThread);
+	return result;
+
+}
+
+/**
+ * static native void setCallSiteTargetNormal(CallSite site, MethodHandle target)
+ */
+void JNICALL
+Java_java_lang_invoke_MethodHandleNatives_setCallSiteTargetNormal(JNIEnv *env, jclass clazz, jobject callsite, jobject target)
+{
+	J9VMThread *currentThread = (J9VMThread*)env;
+	const J9InternalVMFunctions *vmFuncs = currentThread->javaVM->internalVMFunctions;
+	vmFuncs->internalEnterVMFromJNI(currentThread);
+
+	if ((NULL == callsite) || (NULL == target)) {
+		vmFuncs->setCurrentExceptionUTF(currentThread, J9VMCONSTANTPOOL_JAVALANGNULLPOINTEREXCEPTION, NULL);
+	} else {
+		j9object_t callsiteObject = J9_JNI_UNWRAP_REFERENCE(callsite);
+		j9object_t targetObject = J9_JNI_UNWRAP_REFERENCE(target);
+
+		UDATA offset = (UDATA)vmFuncs->instanceFieldOffset(currentThread, J9VM_J9CLASS_FROM_HEAPCLASS(currentThread, callsiteObject), (U_8*)"target", strlen("target"),  (U_8*)"Ljava/lang/invoke/MethodHandle;", strlen("Ljava/lang/invoke/MethodHandle;"), NULL, NULL, 0);
+		MM_ObjectAccessBarrierAPI objectAccessBarrier = MM_ObjectAccessBarrierAPI(currentThread);
+		objectAccessBarrier.inlineMixedObjectStoreObject(currentThread, callsiteObject, offset, targetObject, false);
+	}
+	vmFuncs->internalExitVMToJNI(currentThread);
+}
+
+/**
+ * static native void setCallSiteTargetVolatile(CallSite site, MethodHandle target);
+ */
+void JNICALL
+Java_java_lang_invoke_MethodHandleNatives_setCallSiteTargetVolatile(JNIEnv *env, jclass clazz, jobject callsite, jobject target)
```

Comment:
This API and `Java_java_lang_invoke_MethodHandleNatives_setCallSiteTargetNormal` are identical except for a boolean.  Can you introduce a helper and pass the boolean to it?

Reply:
added common helper